classifier.py

```diff
@@ -3130,59 +3090,436 @@ def target_age(cls, identifier, name):
+        #
+        # TODO: This has not been calibrated.
+        if (audience==Classifier.AUDIENCE_ADULT
+            and adults_only_weight > total_adult_weight/4):
+            audience = Classifier.AUDIENCE_ADULTS_ONLY
+
+        return audience
+
+    def target_age(self, audience):
+        """Derive a target age from the gathered data."""
+
+        if audience not in (
+                Classifier.AUDIENCE_CHILDREN, Classifier.AUDIENCE_YOUNG_ADULT
+        ):
+            # This is not a children's or YA book. Assertions about
+            # target age are irrelevant and the default value rules.
+            return Classifier.default_target_age_for_audience(audience)
+
+        # Only consider the most reliable classifications.
+        reliable_classifications = self.most_reliable_target_age_subset
+
+        # Try to reach consensus on the lower and upper bounds of the
+        # age range.
+        target_age_mins = []
+        target_age_maxes = []
+        for c in reliable_classifications:
+            target_min = c.subject.target_age.lower
+            target_max = c.subject.target_age.upper
+            if target_min is not None:
+                if not c.subject.target_age.lower_inc:
+                    target_min += 1
+                for i in range(0,c.weight):
+                    target_age_mins.append(target_min)
+            if target_max is not None:
+                if not c.subject.target_age.upper_inc:
+                    target_max -= 1
+                for i in range(0,c.weight):
+                    target_age_maxes.append(target_max)
+
+        target_age_min = None
+        target_age_max = None
+        if target_age_mins:
+            # We have opinions about the proper minimum age.
+            mins = Counter(target_age_mins)
+            [(target_age_min, count)] = mins.most_common(1)
+            if count == 1:
+                # Everyone has a different opinion. Pick the smallest
+                # minimum age.
+                target_age_min = min(target_age_mins)
+
+        if target_age_maxes:
+            # We have opinions about the proper maximum age.
+            maxes = Counter(target_age_maxes)
+            [(target_age_max, count)] = maxes.most_common(1)
+            if count == 1:
+                # Everyone has a different opinion. Pick the largest
+                # maximum age.
+                target_age_max = max(target_age_maxes)
+
+        if not target_age_min and not target_age_max:
+            # We found no opinions about target age. Use the default.
+            return Classifier.default_target_age_for_audience(audience)
+
+        if target_age_min is None:
+            target_age_min = target_age_max
+
+        if target_age_max is None:
+            target_age_max = target_age_min
+
+        # If min and max got mixed up somehow, un-mix them.
+        if target_age_min > target_age_max:
+            target_age_min, target_age_max = target_age_max, target_age_min
```

Comment:
That's true, but this should never happen; it's just here as a backstop in case something goes wrong earlier in the process. So I'm not too worried about it.

I did change some of the code that deals with Subjects as they come in, so that a weird subject like "Ages 14-12" will be treated as saying the minimum age is 12 and the maximum age is 14. That's the only situation I can think of where this might have happened. I'd still like to keep the backstop though.